README.md
# Polonium

![Crates.io](https://img.shields.io/crates/v/polonium)

Yet another hacking framework.

# How to install

You can install the latest version of Polonium by running:

```sh
cargo install polonium
```

CHANGELOG.md
# Changelog

## [Unreleased]

## [0.2.0] - 2020-07-27
### Added

### Changed
- [#1](https://github.com/creekorful/polonium/issues/1) Rename executables.
- [#3](https://github.com/creekorful/polonium/issues/3) Allow to read ports from file.
- [#4](https://github.com/creekorful/polonium/issues/4) Ship file with common used ports.

### Fixed
- [#2](https://github.com/creekorful/polonium/issues/2) Now perform hostname resolution.

## [0.1.0] - 2020-07-1O

Initial pre-release.

[unreleased]: https://github.com/creekorful/polonium/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/creekorful/polonium/releases/tag/v0.2.0
[0.1.0]: https://github.com/creekorful/polonium/releases/tag/v0.1.0

src/grabbing.rs
use std::io::{ErrorKind, Read, Write};
use std::net::TcpStream;
use std::time::Duration;
use std::{error, str};

use crate::{resolve, DEFAULT_CONNECT_TIMEOUT, READ_CONNECT_TIMEOUT, WRITE_CONNECT_TIMEOUT};

/// Grab banner of given target.
///
/// # Examples
///
/// Try to grab banner of service running on localhost:80
///
/// ```no_run
/// use polonium::grabbing::grab_banner;
/// let banner = grab_banner("127.0.0.1:80", &None, &None, &None);
/// ```
pub fn grab_banner(
    address: &str,
    connect_timeout: &Option<Duration>,
    read_timeout: &Option<Duration>,
    write_timeout: &Option<Duration>,
) -> Result<String, Box<dyn error::Error>> {
    let address = resolve(address)?;

    let mut stream =
        TcpStream::connect_timeout(&address, connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT))?;
    stream.set_read_timeout(Option::from(read_timeout.unwrap_or(READ_CONNECT_TIMEOUT)))?;
    stream.set_write_timeout(Option::from(write_timeout.unwrap_or(WRITE_CONNECT_TIMEOUT)))?;

    let mut buffer = Vec::new();

    // Try to read banner right after connecting
    let result = stream.read_to_end(&mut buffer);
    if result.is_ok() && !buffer.is_empty() {
        return Ok(String::from_utf8_lossy(&buffer).to_string());
    }

    // If timeout related error happens, do not fails
    // because we may need to talk first
    let error = result.err().unwrap();
    if error.kind() != ErrorKind::WouldBlock {
        return Err(error.into());
    }

    // If nothing was returned, send a dummy request
    stream.write_all("HEAD / HTTP/1.1\n\n".as_ref())?;

    // Try to read again
    stream.read_to_end(&mut buffer)?;
    Ok(String::from_utf8_lossy(&buffer).to_string())
}


src/scanning.rs
use std::error;
use std::net::TcpStream;
use std::time::Duration;

use crate::{resolve, DEFAULT_CONNECT_TIMEOUT};

/// Scan given address for open ports.
///
/// # Examples
///
/// Scan localhost on ports 80, 8080:
///
/// ```no_run
/// use polonium::scanning::scan;
/// use std::net::SocketAddr;
/// let open_ports = scan("127.0.0.1", &[80, 8080], &None);
/// ```
pub fn scan(
    address: &str,
    ports: &[u16],
    connect_timeout: &Option<Duration>,
) -> Result<Vec<u16>, Box<dyn error::Error>> {
    let mut open_ports: Vec<u16> = Vec::new();
    for port in ports.iter() {
        let target = resolve(&format!("{}:{}", address, port))?;
        if TcpStream::connect_timeout(&target, connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT))
            .is_ok()
        {
            open_ports.push(*port)
        }
    }

    Ok(open_ports)
}


src/lib.rs
use std::error;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::net::{SocketAddr, ToSocketAddrs};
use std::time::Duration;

pub mod grabbing;
pub mod scanning;

const DEFAULT_CONNECT_TIMEOUT: Duration = Duration::from_secs(1);
const READ_CONNECT_TIMEOUT: Duration = Duration::from_secs(1);
const WRITE_CONNECT_TIMEOUT: Duration = Duration::from_secs(1);

/// Parse a duration from given milliseconds string.
///
/// # Examples
///
/// ```
/// use polonium::parse_duration;
/// use std::time::Duration;
/// let duration = parse_duration(&Some("42"));
/// assert_eq!(duration, Some(Duration::from_millis(42)));
/// ```
pub fn parse_duration(value: &Option<&str>) -> Option<Duration> {
    value
        .map(|v| v.parse::<u64>())
        .filter(|v| v.is_ok())
        .map(|v| Duration::from_millis(v.unwrap()))
}

/// Perform the DNS resolution of given hostname (if any)
///
/// # Examples
///
/// ```no_run
/// use polonium::resolve;
/// use std::net::{SocketAddrV4, SocketAddr};
/// let addr = resolve("localhost:80");
/// assert_eq!(addr.unwrap(), SocketAddr::from(([127, 0, 0, 1], 80)));
/// ```
pub fn resolve(address: &str) -> Result<SocketAddr, std::io::Error> {
    address.to_socket_addrs().map(|i| i.as_slice()[0])
}

/// Function that parse ports from several formats
///
/// # Examples
///
/// ```
/// use polonium::parse_ports;
/// assert_eq!(parse_ports("80").unwrap(), [80]);
/// assert_eq!(parse_ports("80,443,8080").unwrap(), [80, 443, 8080]);
/// assert_eq!(parse_ports("20-22").unwrap(), [20, 21, 22]);
/// ```
pub fn parse_ports(ports_arg: &str) -> Result<Vec<u16>, Box<dyn error::Error>> {
    let mut ports: Vec<u16> = Vec::new();

    // single port value
    let port = ports_arg.parse::<u16>();
    if let Ok(port) = port {
        ports.push(port);
        return Ok(ports);
    }

    // file given
    if ports_arg.starts_with('@') {
        let path = ports_arg.replace('@', "");

        // read file line by line
        let ports: Vec<u16> = BufReader::new(File::open(path)?)
            .lines()
            .filter(|p| p.is_ok())
            .map(|p| p.unwrap().parse::<u16>().unwrap())
            .collect();
        return Ok(ports);
    }

    // multiple ports given (i.e 80,443,8080)
    if ports_arg.contains(',') {
        ports = ports_arg
            .split(',')
            .map(|p| p.parse::<u16>().unwrap())
            .collect();
        return Ok(ports);
    }

    // range given (i.e 80-3000)
    if ports_arg.contains('-') {
        let parts: Vec<&str> = ports_arg.split('-').collect();
        let start = parts[0].parse::<u16>()?;
        let end = parts[1].parse::<u16>()?;

        for i in start..end + 1 {
            ports.push(i);
        }

        return Ok(ports);
    }

    Err(format!("`{}` is not a valid format.", ports_arg).into())
}
